vignettes/python-primer.Rmd

```diff
@@ -0,0 +1,920 @@
+    `as.logical()` in R, but returning only a scalar, not a vector).
+
+-   `__repr__`, `__str__`, functions invoked for formatting and pretty
+    printing (equivalent to `format()` and `print()` S3 generics in R.
+
+-   `__enter__` and `__exit__`: functions invoked by `with`.
+
+-   Many builtin Python functions are just sugar for invoking the
+    dunder. For example: calling `repr(x)` is identical to
+    `x.__repr__()`. Other builtins that are just sugar for invoking the
+    dunder are `next()`, `iter()`, `str()`, `list()`, `dict()`,
+    `bool()`, `dir()` and more!
+
+#### Iterators, revisited
+
+Now that we have the basics of `class`, it's time to revisit iterators.
+First, some terminology:
+
+**iterable**: something that can be iterated over. Concretely, a class
+that defines an `__iter__` method, whose job is to return an *iterator*.
+
+**iterator**: something that iterates. Concretely, a class that defines
+a `__next__` method, whose job is to return the next element each time
+it is called, and then raises a `StopIteration` exception once it's
+exhausted.
+
+It's common to see classes that are both iterables and iterators, where
+the `__iter__` method is just a stub that returns `self`.
+
+Here is a custom iterable / iterator implementation of Python's `range`
+(similar to `seq` in R)
+
+```{python, error = TRUE}
+class MyRange:
+  def __init__(self, start, end):
+    self.start = start
+    self.end = end
+
+  def __iter__(self):
+    # reset our counter.
+    self._index = self.start - 1
+    return self
+
+  def __next__(self):
+    if self._index < self.end:
+      self._index += 1 # increment
+      return self._index
+    else:
+      raise StopIteration
+
+for x in MyRange(1, 3):
+  print(x)
+
+# doing what `for` does, but manually
+r = MyRange(1, 3)
+it = iter(r)
+next(it)
+next(it)
+next(it)
+next(it)
+```
+
+### Decorators
+
+Decorators are just functions that take a function as an argument and
+then return a function. Any function can be invoked as a decorator with
+the `@` syntax, which is just sugar for this simple action:
+
+```{python}
+def my_decorator(func):
+  func.x = "a decorator modified this function by adding an attribute `x`"
+  return func
+
+def my_function(): pass
+my_function = my_decorator(my_function)
+
+# @ is just fancy syntax for the above two lines
+@my_decorator
+def my_function(): pass
+```
+
+One decorator you might encounter frequently is:
+
+-   `@property`, which automatically calls a class method when the
+    attribute is accessed (similar to `makeActiveBinding()` in R).
+
+### Defining Generators with `yield`.
+
+Generators are special Python functions that contain one or more `yield`
+statements. As soon as `yield` is included in a code block passed to
+`def`, the semantics change substantially. You're no longer defining a
+mere function, but a generator constructor! In turn, calling a generator
+constructor creates a generator object, which is just another type of
```

Comment:
generator object == generator _class_, is that correct?

Reply:
I *think* so. 
My impression is that the terminology gets a little imprecise here the way it's used in the broader world. In many places people seem to interchangeably use the bare word "generator" for both the 'generator constructor' and the 'generator iterator' (which self-identifies as a 'generator object' when printed at the REPL, and `<class 'generator'>` when queried with `type()`).

Reply:
yeah, I see, terminology stuff ... here too, probably preferences are determined by how one was "socialized" ... plus, the terms you use there are pretty clear